[Dependency Analysis] Add Android Gradle Plugin

ParaskP7 opened the pull request:
Project Thread: paaHJt-6yU-p2
Required By: [Buildkite CI#445](https://github.com/Automattic/buildkite-ci/pull/445)

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR adds the dependency analysis Android Gradle plugin to this project, for dependency analysis purposes.

-----

For now, only the main `buildHealth` task is going to be utilized and produce data once every week on CI (see this [commit](https://github.com/woocommerce/woocommerce-android/pull/11722/commits/8ec89afcb54dcef6ee22809928bfd39d4d09312a) and [this](https://github.com/Automattic/buildkite-ci/pull/445) PR). Amongst other, this data will include the following:

- Number of modules (`projectCount`)
- Unused dependencies which should be removed (`unusedCount`)
- Transitive dependencies which should be declared directly (`undeclaredCount`)
- Existing dependencies which should be modified to be as indicated (`misDeclaredCount`)
- Dependencies which could be compile-only (`compileOnlyCount`)
- Dependencies which should be removed or changed to runtime-only (`runtimeOnlyCount`)

Afterward, this data will get collected from CI and uploaded to our Apps Metrics infrastructure, for visualization and alerting purposes.

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

1. Local: Run the `./gradlew buildHealth` task and verify that under the root level `build/reports/dependency-analysis` folder you get the below 2 reports both, in JSON and text format:
    - `build-health-report.json`
    - `build-health-report.txt`
2. CI: Using the `New Build` 🟢 CI button for [WCAndroid](https://buildkite.com/automattic/woocommerce-android/), test this standalone dependency analysis job (see form below). Then, when CI completes, make sure that the below 2 CI artifacts are available both, in JSON and text format:
    - `build/reports/dependency-analysis/build-health-report.json`
    - `build/reports/dependency-analysis/build-health-report.txt`

![image](https://github.com/woocommerce/woocommerce-android/assets/9729923/cf93ec90-7bec-410e-9a61-d6677eb907ac)

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

ParaskP7 committed:
Build: Add dependency analysis plugin

FYI: This commit adds a previous version of the plugin (1.28.0) and not
the latest version (1.32.0). This is because the latest version gets
stuck during the 'computeActualUsage' phase and increases the build time
of the 'buildHealth' task by as much as five times (5x).

For more info see: Stuck in computeActualUsageDebug for many minutes
#1186 (https://github.com/autonomousapps/
dependency-analysis-gradle-plugin/issues/1186)

ParaskP7 committed:
CI: Add dependency analysis job to pipeline

ParaskP7 committed:
Build: Ignored variants configuration for dependency analysis plugin

This commit updates the encrypted 'gradle.properties.enc' file to point
to that pinned hash which has the
'dependency.analysis.android.ignored.variants'
property added into the 'gradle.properties' file for the WCAndroid
project.

FYI: With this configuration only the 'debug' variant (build type) for
the lib modules and the 'jalapenoDebug' variant (product flavor +
build type) for the app module is going to be included, with the rest of
the variants ignored.

dangermattic commented:
<!--
  0 Errors
  1 Warning: PR is not assigned to a milest...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into build/dependency-analysis-android-gradle-plugin-v2

ParaskP7 committed:
Revert "CI: Add dependency analysis job to pipeline"

This reverts commit 025c16ab1d33680c0c9c550bb3e3b30ad82d3b0a.

ParaskP7 committed:
CI: Add a scheduled dependency analysis job

FYI: This job will be then used by 'buildkite-ci' and configured as a
'buildkite_pipeline_schedule' with a weekly frequently.

AliSoftware commented on .buildkite/schedules/dependency-analysis.yml:
```diff
@@ -0,0 +1,18 @@
+agents:
+  queue: "android"
+
+steps:
+  - label: "dependency analysis"
+    command: |
+      echo "--- 📊 Analyzing"
+      cp gradle.properties-example gradle.properties
+      ./gradlew buildHealth
+    plugins: [*ci_toolkit]
+    artifact_paths:
+      - "build/reports/dependency-analysis/build-health-report.*"
```
I wonder if it'd be worth asking Buildkite to notify in a Slack channel if the CI build failed for some reason, otherwise it being a weekly build it'd be easy to miss if it started failing for some reason.

This can be done by adding the following key to this step:

```yaml
      - slack:
          channels:
            - "#woo-android-notifs"
        if: build.state == "failed"
```

ParaskP7 replied:
Oh, this is a great idea @AliSoftware , will do and thanks for the review! 🥇 

I just wonder if this is the most appropriate channel to notify for failures. Maybe using the generic `android-notifications` Slack channel is better since we would anyway start adding this functionality to more and more client, but also, lib repos. Wdyt?

PS: Also, this `woo-android-notifs` is the only one such Slack channel. Apart from the `fluxc-android` channel, I don't see any other product teams have their own `android-notifs` related channel, and maybe, it is not worth it create a channel for each product, just like me, who didn't know its existence, I am sure others don't as well, which might also include the Android folks working for Woo? 🤔

AliSoftware replied:
Yeah tbh I didn't know about that channel either, all I did was go into Slack, hit ⌘K, and type "woo and" to see if there were a Slack channel for WooCommerce Android in general, and that's when I discovered that one 😛 

Just pick whatever Slack channel would make sense for you and the team and that specific project.

ParaskP7 replied:
> Just pick whatever Slack channel would make sense for you and the team and that specific project.

`android-notifications` Slack channel it is then, once more, many thanks for the recommendation! 🙇

ParaskP7 replied:
Btw @AliSoftware are you sure the [above](https://github.com/woocommerce/woocommerce-android/pull/11722#discussion_r1639902355) is the right way of doing that, and not doing this below instead ([docs](https://buildkite.com/docs/pipelines/notifications) + [example](https://github.com/woocommerce/woocommerce-android/blob/trunk/.buildkite/release-builds.yml#L24-L25)):

```yml
    notify:
      - slack: "#android-notifications"
        if: build.state == "failed"
```

🤔

AliSoftware replied:
> Btw @AliSoftware are you sure the https://github.com/woocommerce/woocommerce-android/pull/11722#discussion_r1639902355 is the right way of doing that, and not doing this below instead ([docs](https://buildkite.com/docs/pipelines/notifications) + [example](https://github.com/woocommerce/woocommerce-android/blob/trunk/.buildkite/release-builds.yml#L24-L25)):

Both are correct… but yours is actually better 🙂 

 - `slack: "#channelname"` is just a shorthand for `slack: { channels: ["#channelname"] }`, for the common case where you only want to notify in a single channel, and don't want to customize anything else of the Slack message. [If you wanted to also customize the message that Buildkite sends on failure, you'd need to also add a `message` subkey under `slack`](https://buildkite.com/docs/pipelines/notifications#slack-channel-and-direct-messages-custom-messages), so that's when you can't just use the `slack: "#channelname"` shorthand and have to start using the expanded form `slack: { channels: ["#channelname"], message: "Oops" }`. And similarly if you wanted to notify _multiple_ Slack channels the shorthand form wouldn't be enough either. But in our case when we just need to provide a single channel name and no custom message, they are equivalent—and your form is shorter so probably nicer 🙂 
 - Another thing that we could vary is where we put that `notify` key. In my comment above I suggested to add the `notify` key to the `step`, which means that Buildkite will only report in Slack if _that particular step/job_ failed. But you can also add the `notify` key at the root level of the YAML instead, to ask Buildkite to notify if _any_ of the `steps`/jobs in this pipeline fail. In our specific case because you only have one `step` in that pipeline that won't make a difference, but if later you add more `steps` to this `.yml` then it probably makes more sense to ask to be notified in Slack if _any_ of those would fail. So even if for the current setup it doesn't make a difference, conceptually it would probably make more sense to put that `notify` at the root of the pipeline after all…

ParaskP7 replied:
Thanks for the clarification on all that @AliSoftware ! 🙇 

I don't foresee us adding any more steps to this pipeline, so I actually think that keeping it simple and targeted for now is better, hope you agree, I am pushing... 🫂

AliSoftware reviewed (approved):
> Then, when CI completes, make sure that the below 2 CI artifacts are available both, in JSON and text format:
> - `build/reports/dependency-analysis/build-health-report.json`
> - `build/reports/dependency-analysis/build-health-report.txt`

I didn't trigger a manual build myself and instead looked at the one you triggered on your end, and can confirm the artifacts were there 🙂 

<img width="943" alt="image" src="https://github.com/woocommerce/woocommerce-android/assets/216089/cc281159-6559-47c8-a448-e0eb3f829e80">

ParaskP7 committed:
CI: Notify on the android-notifications slack channel on a build failure

wzieba commented on build.gradle:
```diff
@@ -5,6 +5,7 @@ import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
 plugins {
     id 'io.gitlab.arturbosch.detekt'
     id 'com.automattic.android.measure-builds'
+    id "com.autonomousapps.dependency-analysis"
```
Have you been considering applying the plugin as [initialization script](https://docs.gradle.org/current/userguide/init_scripts.html)? This plugin will be only used on CI, so it'd be the best if we saved configuration time for local builds by not applying the plugin on local machines.

I've used this technique in Autostattic ([example](https://github.com/wzieba/Autostattic/blob/main/script/add-versions-plugin.gradle)) and it worked pretty well.

ParaskP7 replied:
Hmmmm, I haven't though about it, thanks for the suggestion @wzieba ! 🥇 

> This plugin will be only used on CI, so it'd be the best if we saved configuration time for local builds by not applying the plugin on local machines.

I agree, let me check how would that work, I haven't tried something similar in the past so I don't know what to expect. 🤔 

> I've used this technique in Autostattic ([example](https://github.com/wzieba/Autostattic/blob/main/script/add-versions-plugin.gradle)) and it worked pretty well.

Thanks for sharing, let me try figuring out how would that be applied to a normal project such as WCAndroid, since looking at Autostattic I am seeing a different Grade configuration to what we are used to (in general).

-----

Btw, do you think the configuration time will increase significantly with that plugin applied, and if so, how much (approximately)? 🤔 

Ideally, I would also want anyone to be able to trigger this plugin's task locally, that is, whenever they might need to use this tool, to just experimenting with it or even more. I am only asking that because I am not sure if this change you suggestion, and I am about to make, will make it harder to run this tool locally.

wzieba replied:
No, the increase won't be significant, but the plugin will have to be configured for each build. We're talking range of hundreds of ms per build I'd say.

While dependency analysis plugin is not a problem on its own, additional build overhead will add up. E.g. I don't think it's great that we apply `se.bjurr.violations.violation-comments-to-github-gradle-plugin` plugin, as it has completely no usage in local environment. 

> Ideally, I would also want anyone to be able to trigger this plugin's task locally

I thought we want to use this plugin only on CI. If local usage is the case as well, then please ignore my suggestion - initialization script won't have usage here.

ParaskP7 replied:
> I thought we want to use this plugin only on CI. If local usage is the case as well, then please ignore my suggestion - initialization script won't have usage here.

Yea, I would like us to be able to use this plugin locally as well. It has so much more potential than just running `buildHealth` at the root level for us to get a holistic view on what-is-what and report that weekly, which is the task we want to run on CI.

FYI: You could check this [Tasks](https://github.com/autonomousapps/dependency-analysis-gradle-plugin/wiki/Tasks) documentation to understand more about what this plugin can offer. For example, check this [Reason](https://github.com/autonomousapps/dependency-analysis-gradle-plugin/wiki/Tasks#reason) task. I expect devs to use this and other such tasks ([Fix Dependencies](https://github.com/autonomousapps/dependency-analysis-gradle-plugin/wiki/Tasks#fix-dependencies)) locally to try and make the build "healthier". 🤞

wzieba replied:
> Btw, do you think the configuration time will increase significantly with that plugin applied, and if so, how much (approximately)? 🤔

On clean build, fetching dependencies for this plugin took around 22s of serial time (**not** wall clock time, as fetching is parallelized): https://scans.gradle.com/s/pmec7ax2v54ro/performance/network-activity

Configuration time (with fetched dependencies) is around 170ms for each build, but it's rather not noticeable in the context of ~3m build.

ParaskP7 replied:
Awesome, thanks so much for this breakdown! 🙇

wzieba commented on gradle.properties-example:
```diff
@@ -56,3 +56,6 @@ tracksEnabled = false
 android.nonTransitiveRClass=true
 android.enableR8.fullMode=false
+
+# Dependency Analysis Plugin
+dependency.analysis.android.ignored.variants=release,vanillaDebug,vanillaRelease,wasabiDebug,wasabiRelease,jalapenoRelease
```
I think it'd be awesome if we found a way to apply this setting not in `gradle.properties` but somewhere in `build.gradle`. Otherwise, for each change we'll have to keep our secret file with this example one in sync.

ParaskP7 replied:
> Otherwise, for each change we'll have to keep our secret file with this example one in sync.

What do you mean by that, can you elaborate a bit more? 🤔 

FYI: I don't expect any change we do thereafter to affect this configuration going forward. This is configuration purely makes this plugin focus on working with a single variant. Unless of course we remove, update, or add, build variants. But, I haven't seen that happening too often on any of the projects I've worked with so far. 🤷 

> I think it'd be awesome if we found a way to apply this setting not in gradle.properties but somewhere in build.gradle.

I guess so, but also, I don't want to make this more too complicated for local usage. I am guess you referring to what we've discussed few weeks ago on using environmental variables when I posted this RFC on ignored variants, right? 🤔

wzieba replied:
> What do you mean by that, can you elaborate a bit more? 🤔

If we edit value of `dependency.analysis.android.ignored.variants` in `gradle.properties-example`, AFAIU we should update this value in the secret `gradle.properties` as well.

> I guess so, but also, I don't want to make this more too complicated for local usage. I am guess you referring to what we've discussed few weeks ago on using environmental variables when I posted this RFC on ignored variants, right? 🤔

No, I was looking for a way to define this option in `build.gradle`, to have a single source. I don't know why authors of the plugin didn't expose an option in the plugin extension, but there's probably a reason for that. Ideally, what I'd like to have is

```
dependencyAnalysis {
    ignoredVariants.set("release", "vanillaDebug" //etc.)
}
```

somewhere in `build.gradle`. As it's not possible, an alternative I see is to set `dependency.analysis.android.ignored.variants` property in `build.gradle`, rather than setting this property 2 times - one time in example `gradle.properties` and again in secrets one. It's not a blocker, though.

ParaskP7 replied:
> If we edit value of dependency.analysis.android.ignored.variants in gradle.properties-example, AFAIU we should update this value in the secret gradle.properties as well.

Yea, I understand that, as this is exactly what I did in this PR.

> No, I was looking for a way to define this option in build.gradle, to have a single source. I don't know why authors of the plugin didn't expose an option in the plugin extension, but there's probably a reason for that. Ideally, what I'd like to have is somewhere in build.gradle. As it's not possible, an alternative I see is to set dependency.analysis.android.ignored.variants property in build.gradle

Yea, unfortunately, this is not an available option.

> rather than setting this property 2 times - one time in example gradle.properties and again in secrets one. It's not a blocker, though.

Yea, I understand that, but as I explained, I am envisioning (and honestly hoping) that we will not need to change this flag any more, just once during the initial setup of this plugin on any repo, for it to ignore all variant and only have the `debug` build type + `dev` product flavor variant enabled when working with this plugin.

Another way of working with this plugin is to have the `release` + `production` variant enabled instead, which I also considered, just so we don't change our minds in the future. But, after thinking about it, I decided to go with the `debug` build type + `dev` product flavor variant because we do use a lot more `debug` dependencies comparing to `release` ones, and those `debug` tend to change more often that the `release` dependencies, thus I though it is better to focus on that variant going forward. Plus, `debug` build type + `dev` product flavor variant is faster, although, us running it once a week, we no longer care about that.

Now, us having decided on which single variant to use, the `debug` build type + `dev` VS. `release` + `production` variant, and then configuring it, should be the end of it, at least on CI. For local builds, devs could then go on and flip that flag according to their needs, for testing purposes and beyond.

wzieba reviewed (commented):
I've commented with an idea of initialization script - WDYT?

ParaskP7 commented on .buildkite/schedules/dependency-analysis.yml:
```diff
@@ -0,0 +1,21 @@
+    artifact_paths:
+      - "build/reports/dependency-analysis/build-health-report.*"
+    notify:
+      - slack: "#android-notifications"
```
👋 @wzieba !

One thing I am still not sure about is if this `android-notifications` channel is the best one to use. Wdyt? 🤔 

To my understanding, although this channel, its name and probable its purpose is generic, it is only used for JP/WPAndroid related purposes, right? 🤔 

Then, we have this `woo-android-notifs` and `fluxc-android` + `fluxc-notifications`, which have a non-generic name, and focused on a specific product/lib, just like the `android-notifications` channel does for JP/WPAndroid.

Ideally, I would like us to have one Slack notification channel for all such build failures, not a specific separate channel per product, but then again, maybe this is not the right more, not sure, what are your thoughts? 🙏

wzieba replied:
I think for WooCommerce Android we should use `woo-android-notifs` and not `android-notifications` as the later, as you correctly noticed, is JP/WP related. 

I'm also okay to have a channel like `android-core-notifs` which would collect all alerts that _we should act upon_ so we don't have to jump between all channels.

ParaskP7 replied:
Yea, thanks for the `android-core-notifs` suggestion @wzieba , I think for this case, this is exactly what we need, as this is about the task failing on CI, for unknown build related reasons, not that the task is failing due to dependency related problems that product devs should fix. Also, only JP/WP/WCAndroid are having these channels, I am not seeing any other product team using such channels, right? 🤔 

The later is just the next and "future" step, when we'll get that flow working with Apps Metrics and product team start to (hopeful) fully utilize this tool and data. And this later flow needs further configuration and an initial clean-up of the dependencies anyway, like remove all unused dependencies first and then make the plugin fail on any new unused dependency it finds. No matter, something for us to think later on.

So, let's create and use this `android-core-notifs` for starters, right?

ParaskP7 replied:
FYI: Taking your 👍 on my creating the Slack channel, I went ahead and made this `android-core-notifs` change.

wzieba replied:
We discussed this internally: p1718625356458629-slack-C0821FQTA

ParaskP7 committed:
CI: Update android-notifications to android-core-notifs on CI failures

More more info see: https://github.com/woocommerce/woocommerce-android/
pull/11722#discussion_r1642530847

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into build/dependency-analysis-android-gradle-plugin-v2
